ConfigurationBindingGenerator Fails with ArgumentOutOfRangeException

### Description

When configuration binding generation is enabled by means of `<EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>`, no source code is generated. Instead, the generator throws an `ArgumentOutOfRangeException` and fails with this message:

```
CSC : warning CS8785: Generator 'ConfigurationBindingGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of 
type 'ArgumentOutOfRangeException' with message 'repeatCount ('-4') must be a non-negative value. (Parameter 'repeatCount') [/app/repro/repro.csproj]
```

### Reproduction Steps

Given the following project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0-preview.4.23259.5" />
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0-preview.4.23259.5" />
  </ItemGroup>

</Project>
```

…and the following Program.cs:

```c#
using Microsoft.Extensions.Configuration;

var c = new ConfigurationBuilder().Build();
var x = c.GetSection("Anything").Get<TimeSpan?>();
Console.WriteLine(x.GetValueOrDefault());
```

…run `dotnet build` via public 8.0 Preview 4 or daily 8.0 Preview 5.

This repro was generated from the `dotnet new console` command. The method `Bind` can be substituted with the same result. 8.0 Preview 5 versions of "Microsoft.Extensions.Configuration.Binder" and "Microsoft.Extensions.Hosting" can be substituted with the same result.

(Interestingly, source generation isn't even _attempted_ unless "Microsoft.Extensions.Hosting" is present. This was also the case in 8.0 Preview 3. Is that meant to be so? I couldn't find it documented. If not, I can file another issue.)

### Expected behavior

A file is generated which contains configuration binding code.

### Actual behavior

Compilation fails with the following error:

```
CSC : warning CS8785: Generator 'ConfigurationBindingGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of 
type 'ArgumentOutOfRangeException' with message 'repeatCount ('-4') must be a non-negative value. (Parameter 'repeatCount') [/app/repro/repro.csproj]
```

### Regression?

This was not present in 8.0 Preview 3.

### Configuration

.NET Version: This has been tried with "8.0.100-preview.4.23260.5" and "8.0.100-preview.5.23269.37" versions of the SDK.

OS and version: macOS 13.3.1 (a) and Ubuntu 18.04.6 LTS. The latter via `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-arm64-20220427171722-6e40d49`. (~Though I see that the cross-compilation instructions have changed since I last checked. I'll make sure to try to recreate with `mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-arm64`.~ _Same on "CBL-Mariner 2.0.20230426"._)

Architecture: x64. I don't believe it is specific to that configuration.

Blazor: :x:

### Other information

I think this has something to do with indentation determination. The only `repeatCount` it seems that it could plausibly be is the one in `StringBuilder`'s `AppendChar`.

> could you please try referencing the following package […]

Sure thing:

```diff
     <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0-preview.4.23259.5" />
+    <PackageReference Include="Microsoft.Net.Compilers.Toolset" Version="4.7.0-1.final" />
   </ItemGroup>
```

But no difference.

```
root [ /app/repro ]# ~/.dotnet/dotnet build
MSBuild version 17.7.0-preview-23267-02+fe79798bd for .NET
  Determining projects to restore...
  Restored /app/repro/repro.csproj (in 238 ms).
/root/.dotnet/sdk/8.0.100-preview.5.23269.37/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/app/repro/repro.csproj]
CSC : warning CS8785: Generator 'ConfigurationBindingSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentOutOfRangeException' with message 'repeatCount ('-4') must be a non-negative value. (Parameter 'repeatCount') [/app/repro/repro.csproj]
  repro -> /app/repro/bin/Release/net8.0/repro.dll

Build succeeded.

CSC : warning CS8785: Generator 'ConfigurationBindingSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentOutOfRangeException' with message 'repeatCount ('-4') must be a non-negative value. (Parameter 'repeatCount') [/app/repro/repro.csproj]
    1 Warning(s)
    0 Error(s)
```

With these dependencies:

```xml
<PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0-rc.1.23370.5" />
<PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0-rc.1.23374.2" />
```

…and compiler version `8.0.100-preview.6.23330.14`, I get this error:

```
CSC : warning CS8785: Generator 'ConfigurationBindingGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'InvalidCastException' with message 'Unable to cast object of type 'Microsoft.Extensions.Configuration.Binder.SourceGeneration.NullableSpec' to type 'Microsoft.Extensions.Configuration.Binder.SourceGeneration.ObjectSpec'.' [/Users/cosborn/Developer/repro/repro.csproj]
```

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/CoreBindingHelper.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/CoreBindingHelper.cs
@@ -98,11 +98,12 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
 
                 foreach (TypeSpec type in types)
                 {
-                    TypeSpecKind kind = type.SpecKind;
+                    TypeSpec effectiveType = type.EffectiveType;
+                    TypeSpecKind kind = effectiveType.SpecKind;
 
                     EmitStartBlock($"if (type == typeof({type.MinimalDisplayString}))");
 
-                    if (type is ParsableFromStringSpec stringParsableType)
+                    if (effectiveType is ParsableFromStringSpec stringParsableType)
                     {
                         EmitCastToIConfigurationSection();
                         EmitBindLogicFromString(
@@ -113,9 +114,9 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
                             checkForNullSectionValue: stringParsableType.StringParsableTypeKind is not StringParsableTypeKind.AssignFromSectionValue,
                             useIncrementalStringValueIdentifier: false);
                     }
-                    else if (!EmitInitException(type))
+                    else if (!EmitInitException(effectiveType))
                     {
-                        EmitBindCoreCall(type, Identifier.obj, Identifier.configuration, InitializationKind.Declaration);
+                        EmitBindCoreCall(effectiveType, Identifier.obj, Identifier.configuration, InitializationKind.Declaration);
                         _writer.WriteLine($"return {Identifier.obj};");
                     }
 

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
@@ -217,7 +217,7 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
         }
 
         [Fact]
-        public void GetScalar()
+        public void Get_Scalar()
         {
             var dic = new Dictionary<string, string>
             {
@@ -239,7 +239,7 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
         }
 
         [Fact]
-        public void GetScalarNullable()
+        public void Get_ScalarNullable()
         {
             var dic = new Dictionary<string, string>
             {
@@ -258,6 +258,50 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
             Assert.True((bool)config.GetValue(typeof(bool?), "Boolean"));
             Assert.Equal(-2, (int)config.GetValue(typeof(int?), "Integer"));
             Assert.Equal(11, (int)config.GetValue(typeof(int?), "Nested:Integer"));
+        }
+
+        [Fact]
+        public void GetValue_Scalar()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Integer", "-2"},
+                {"Boolean", "TRUe"},
+                {"Nested:Integer", "11"}
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            Assert.True(config.GetSection("Boolean").Get<bool>());
+            Assert.Equal(-2, config.GetSection("Integer").Get<int>());
+            Assert.Equal(11, config.GetSection("Nested:Integer").Get<int>());
+
+            Assert.True((bool)config.GetSection("Boolean").Get(typeof(bool)));
+            Assert.Equal(-2, (int)config.GetSection("Integer").Get(typeof(int)));
+            Assert.Equal(11, (int)config.GetSection("Nested:Integer").Get(typeof(int)));
+        }
+
+        [Fact]
+        public void GetValue_ScalarNullable()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Integer", "-2"},
+                {"Boolean", "TRUe"},
+                {"Nested:Integer", "11"}
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            Assert.True(config.GetSection("Boolean").Get<bool?>());
+            Assert.Equal(-2, config.GetSection("Integer").Get<int?>());
+            Assert.Equal(11, config.GetSection("Nested:Integer").Get<int?>());
+
+            Assert.True(config.GetSection("Boolean").Get(typeof(bool?)) is true);
+            Assert.Equal(-2, (int)config.GetSection("Integer").Get(typeof(int?)));
+            Assert.Equal(11, (int)config.GetSection("Nested:Integer").Get(typeof(int?)));
         }
 
         [Fact]
@@ -1358,18 +1402,35 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
         [Fact]
         public void CanBindRecordStructOptions()
         {
-            var dic = new Dictionary<string, string>
-            {
-                {"Length", "42"},
-                {"Color", "Green"},
-            };
-            var configurationBuilder = new ConfigurationBuilder();
-            configurationBuilder.AddInMemoryCollection(dic);
-            var config = configurationBuilder.Build();
+            IConfiguration config = GetConfiguration("Length", "Color");
+            Validate(config.Get<RecordStructTypeOptions>());
+            Validate(config.Get<RecordStructTypeOptions?>().Value);
 
-            var options = config.Get<RecordStructTypeOptions>();
-            Assert.Equal(42, options.Length);
-            Assert.Equal("Green", options.Color);
+            config = GetConfiguration("Options.Length", "Options.Color");
+            // GetValue works for only primitives.
+            //Reflection impl handles them by honoring `TypeConverter` only.
+            // Source-gen supports based on an allow-list.
+            Assert.Equal(default(RecordStructTypeOptions), config.GetValue<RecordStructTypeOptions>("Options"));
+            Assert.False(config.GetValue<RecordStructTypeOptions?>("Options").HasValue);
+
+            static void Validate(RecordStructTypeOptions options)
+            {
+                Assert.Equal(42, options.Length);
+                Assert.Equal("Green", options.Color);
+            }
+
+            static IConfiguration GetConfiguration(string key1, string key2)
+            {
+                var dic = new Dictionary<string, string>
+                {
+                    { key1, "42" },
+                    { key2, "Green" },
+                };
+
+                var configurationBuilder = new ConfigurationBuilder();
+                configurationBuilder.AddInMemoryCollection(dic);
+                return configurationBuilder.Build();
+            }
         }
 
         [Fact]

